tensorly/decomposition/_nn_cp.py

```diff
@@ -0,0 +1,413 @@
+
+    return kt
+
+
+def non_negative_parafac(tensor, rank, n_iter_max=100, init='svd', svd='numpy_svd',
+                         tol=10e-7, random_state=None, verbose=0, normalize_factors=False,
+                         return_errors=False, mask=None, orthogonalise=False, cvg_criterion='abs_rec_error',
+                         fixed_modes=[]):
+    """
+    Non-negative CP decomposition
+
+    Uses multiplicative updates, see [2]_
+
+    This is the same as parafac(non_negative=True).
+
+    Parameters
+    ----------
+    tensor : ndarray
+    rank   : int
+            number of components
+    n_iter_max : int
+                 maximum number of iteration
+    init : {'svd', 'random'}, optional
+    svd : str, default is 'numpy_svd'
+        function to use to compute the SVD, acceptable values in tensorly.SVD_FUNS
+    tol : float, optional
+          tolerance: the algorithm stops when the variation in
+          the reconstruction error is less than the tolerance
+    random_state : {None, int, np.random.RandomState}
+    verbose : int, optional
+        level of verbosity
+    fixed_modes : list, default is []
+        A list of modes for which the initial value is not modified.
+        The last mode cannot be fixed due to error computation.
+
+    Returns
+    -------
+    factors : ndarray list
+            list of positive factors of the CP decomposition
+            element `i` is of shape ``(tensor.shape[i], rank)``
+
+    References
+    ----------
+    .. [2] Amnon Shashua and Tamir Hazan,
+       "Non-negative tensor factorization with applications to statistics and computer vision",
+       In Proceedings of the International Conference on Machine Learning (ICML),
+       pp 792-799, ICML, 2005
+    """
+    epsilon = 10e-12
+    rank = validate_cp_rank(tl.shape(tensor), rank=rank)
+
+    if mask is not None and init == "svd":
+        message = "Masking occurs after initialization. Therefore, random initialization is recommended."
+        warnings.warn(message, Warning)
+
+    if orthogonalise and not isinstance(orthogonalise, int):
+        orthogonalise = n_iter_max
+
+    weights, factors = initialize_nn_cp(tensor, rank, init=init, svd=svd,
+                                 random_state=random_state,
+                                 normalize_factors=normalize_factors)
+    rec_errors = []
+    norm_tensor = tl.norm(tensor, 2)
+
+    if tl.ndim(tensor)-1 in fixed_modes:
+        warnings.warn('You asked for fixing the last mode, which is not supported while tol is fixed.\n The last mode will not be fixed. Consider using tl.moveaxis()')
+        fixed_modes.remove(tl.ndim(tensor)-1)
+    modes_list = [mode for mode in range(tl.ndim(tensor)) if mode not in fixed_modes]
+
+    for iteration in range(n_iter_max):
+        if orthogonalise and iteration <= orthogonalise:
+            for i, f in enumerate(factors):
+                if min(tl.shape(f)) >= rank:
+                    factors[i] = tl.abs(tl.qr(f)[0])
```

Comment:
This seems like a really bad approach to orthogonalization. An alternative might be [pivoting](http://proceedings.mlr.press/v97/zhang19r/zhang19r.pdf)? I'd suggest removing it for now...

Reply:
This orthogonalization was taken from the paper "[Orthgonalized ALS for Tensor Decomposition](http://proceedings.mlr.press/v70/sharan17a/sharan17a.pdf)", Sharan and Valiant, ICML 2017.

It was added to the original CP and was never intended to really be used with the non-negative version and might be problematic indeed. Pivoting would be great to have instead! Happy to remove in the meantime.